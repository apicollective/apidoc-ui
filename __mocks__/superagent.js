/* let mockDelay;*/
let mockError;
let mockHeaders = {};
let mockResponses = [];
  /* status: () => 200,
   * ok: true,
   * get: jest.genMockFunction(),
   * toError: jest.genMockFunction(),
   * headers: {},
     };*/
let currentPath = '';

const Request = {
  post(path) {
    currentPath = path;
    return this;
  },
  get(path) {
    currentPath = path;
    return this;
  },
  send: jest.genMockFunction().mockReturnThis(),
  query: jest.genMockFunction().mockReturnThis(),
  field: jest.genMockFunction().mockReturnThis(),
  set: jest.genMockFunction().mockReturnThis(),
  accept: jest.genMockFunction().mockReturnThis(),
  timeout: jest.genMockFunction().mockReturnThis(),
  then: callback => new Promise((resolve, reject) => {
    if (mockError) reject(mockError);
    else resolve(callback(mockResponses[currentPath]));
  }),
  __setMockResponse:  (path, mockRes) => { // eslint-disable-line
    if (!mockRes.headers) mockRes.headers = mockHeaders;
    mockResponses[path] = mockRes;
  },
  __setMockError: (mockErr) => { // eslint-disable-line
    mockError = mockErr;
  },
  __setMockDefaultMockHeaders: (headers) => { // eslint-disable-line
    mockHeaders = headers;
  },
  __reset: () => { mockResponses = []; },
};


module.exports = Request;
