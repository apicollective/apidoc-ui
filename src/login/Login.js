// @flow
import React from 'react';
import { Redirect } from 'react-router';
import type { Location } from 'react-router';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import queryString from 'query-string';

import { actions } from 'login/sagas';
import type { State } from 'login/reducers';

export type Session = {|
  id: string,
|};

type Props = {|
  loggingIn: boolean,
  session?: Session,
  actions?: typeof actions, // Not sure this is accurate
  location?: Location,
|};

/*
   - Goto github.com?...
   - Return to <host>/login?code=abc
   - Display loading
   - Backend call to validate token
   - Once validated
   -- update User, jwt
   -- redirect to home page
 */

const Login = (props: Props) => {
  if (props.loggingIn) {
    return LoggingIn();
  }

  // If we have a valid JWT redirect to /
  if (props.session && props.session.id) {
    // TODO - valide JWT
    return <Redirect push={true} to="/" />;
  }

  // If we've returned from github with a token, validate it with the backend
  if (props.location) {
    const query = queryString.parse(props.location.search);
    if (query.code && props.actions) {
      props.actions.login(query.code);
      return LoggingIn();
    }
  }

  // Not logged in
  return LoginPlease();
};

const LoginPlease = () => <div>Login Please</div>;

const LoggingIn = () => <div>Logging In</div>;

const mapStateToProps = (state: State) => ({
  loggingIn: state.loggingIn,
  session: state.session,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
