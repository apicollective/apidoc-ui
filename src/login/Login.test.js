// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import createHistory from 'history/createBrowserHistory';
import { configureStore } from 'store/configureStore';
import * as superagent from 'superagent';

import Login from 'login/Login';

export const history = createHistory();
const store = configureStore(history);

test('Login Please', () => {
  const props = {};

  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter><Login {...props} /></MemoryRouter>
    </Provider>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Validate github token with backend successfully', done => {
  superagent.__reset();
  const response = {
    body: {
      session: { id: 'session id' },
    },
  };
  superagent.__setMockResponse('/authenticate_github', response);

  const props = {
    location: {
      search: '?code=abc',
    },
  };

  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter><Login {...props} /></MemoryRouter>
    </Provider>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  /* expect(component.toJSON()).toMatchSnapshot();*/

  setTimeout(() => {
    const session = store.getState().login.session || {};
    expect(session.id).toBe('session id');
    done();
  }, 100);
});

test('After backend validation, should redirect back to home page', () => {
  // FIXME
});
