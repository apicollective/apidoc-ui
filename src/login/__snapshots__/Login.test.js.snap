// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Please 1`] = `
<div>
  Login Please
</div>
`;

exports[`Validate github token with backend successfully 1`] = `
<div>
  Logging In
</div>
`;
